controller/src/main/java/io/pravega/controller/server/ControllerServiceMain.java

```diff
@@ -101,34 +102,36 @@ protected void triggerShutdown() {
     protected void run() throws Exception {
         long traceId = LoggerHelpers.traceEnter(log, this.objectId, "run");
         try {
-            if (serviceConfig.isControllerClusterListenerEnabled()) {
+            while (isRunning()) {
+                // Create store client.
+                log.info("Creating store client");
+                storeClient = StoreClientFactory.createStoreClient(serviceConfig.getStoreClientConfig());
 
-                // Controller cluster feature is enabled.
-                while (isRunning()) {
-                    // Create store client.
-                    log.info("Creating store client");
-                    storeClient = StoreClientFactory.createStoreClient(serviceConfig.getStoreClientConfig());
+                boolean hasZkConnection = serviceConfig.getStoreClientConfig().getStoreType().equals(StoreType.Zookeeper) ||
+                        serviceConfig.isControllerClusterListenerEnabled();
 
-                    // Client should be ZK if controller cluster is enabled.
+                CompletableFuture<Void> sessionExpiryFuture = new CompletableFuture<>();
```

Comment:
I'm not entirely sure about this logic. What if the zk client closes normally (not a session expired)? I think we also want to shut down the controller instance, no?

Reply:
i have not changed the existing logic. Merely consolidated the code across if (clusterListenerEnabled) block and ensured we look for session expiration notification if storeclient is of type Zookeeper. 

Also, secondly, we never close zookeeper client except during shutdown. ControllerServiceMain also waits on serviceshutdown signal.